rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -176,6 +177,26 @@ class AnySubscriptionCallback
     }
   }
 
+  void dispatch_intra_process(
+    ConstMessageSharedPtr & message, const rmw_message_info_t & message_info)
+  {
+    (void)message_info;
+    if (const_shared_ptr_callback_) {
+      const_shared_ptr_callback_(std::move(message));
+    } else if (const_shared_ptr_with_info_callback_) {
+      const_shared_ptr_with_info_callback_(std::move(message), message_info);
+    } else {
+      if (unique_ptr_callback_ || unique_ptr_with_info_callback_ ||
```

Comment:
In the case of storing a `ConstMessageSharedPtr` and dispatching to `unique_ptr`, would it make sense to do the copy to a unique ptr here?

Reply:
The signature of `dispatch_intra_process` method depends on the signature of the callback being used, and not on the message being stored.
https://github.com/ros2/rclcpp/blob/85c50717b6ca427f483c60b3129d1f0596b9869c/rclcpp/include/rclcpp/subscription.hpp#L148-L188
e.g.: The signature of the callback is `unique_ptr` and the message stored is a `ConstMessageSharedPtr`.
- `handle_intra_process_message` is call from an executor.
- L159 `any_callback_.use_take_shared_method()` is `false`.
- L175 `take_intra_process_message` subscriber method is called, with `unique_ptr` callback.
https://github.com/ros2/rclcpp/blob/85c50717b6ca427f483c60b3129d1f0596b9869c/rclcpp/include/rclcpp/subscription.hpp#L201-L215
- L213, IPM `take_intra_process_message` is called.
https://github.com/ros2/rclcpp/blob/85c50717b6ca427f483c60b3129d1f0596b9869c/rclcpp/include/rclcpp/intra_process_manager.hpp#L327-L361
- This calls `MappedRingBuffer` `get` or `pop` method with `unique_ptr` signature.
https://github.com/ros2/rclcpp/blob/85c50717b6ca427f483c60b3129d1f0596b9869c/rclcpp/include/rclcpp/mapped_ring_buffer.hpp#L104-L130
- Both of them copy the `ConstMessageSharedPtr` in a `unique_ptr`. e.g. for the `get` method, mapped_ring_buffer.hpp L113 `it->unique_value` is nullptr, so the else clause is executed, which copies the shared_pointer to a unique_ptr.